### ISSUE TITLE
sol2 overridable_function_members suggest!!

### ISSUE THREAD

**Issue by brinkqiang** [ThePhD/sol2#1227]
In the example overridable_function_members. The usage of injection is introduced. I want to ask if there is a non-injection.

Example: This way of use
```
#define SOL_ALL_SAFETIES_ON 1
#include <sol/sol.hpp>

#include <iostream>

int main() {
std::cout << "=== override-able member functions ===" << std::endl;

struct thingy {
thingy(){
}

virtual void paint() {
std::cout << "p" << std::endl;
}
};

sol::state lua;
lua.open_libraries(sol::lib::base);

lua.new_usertype<thingy>("thingy",
sol::constructors<thingy()>(),
"paint", &thingy::paint);

sol::string_view code = R"(
obj = thingy.new()
obj:paint()
obj.paint = function (self) print("g") end
obj:paint()
function obj:paint () print("s") end
obj:paint()
)";

lua.safe_script(code);

std::cout << std::endl;

return 0;
}
```
The demand is actually the hope that the virtual function of the corresponding C++ class can be implemented in Lua.
Now there is no error when writing this way, but all the calls are on the C++ side.

**Comment by ThePhD** [comment]
So one thing to clarify: what do you expect this code to print?

```cpp
#define SOL_ALL_SAFETIES_ON 1
#include <sol/sol.hpp>

#include <iostream>

int main() {
     std::cout << "=== override-able member functions ===" << std::endl;

     struct thingy {
          thingy(){
          }

          virtual void paint() {
               std::cout << "p" << std::endl;
          }
     };

     sol::state lua;
     lua.open_libraries(sol::lib::base);

     lua.new_usertype<thingy>("thingy",
          sol::constructors<thingy()>(),
          "paint", &thingy::paint);

     sol::string_view code = R"(
          obj = thingy.new()
          obj:paint()
          obj.paint = function (self) print("g") end
          obj:paint()
          function obj:paint () print("s") end
          obj:paint()
     )";
     
     /// (!!!!) THIS PART HERE
     thingy* obj_from_lua = lua["obj"];
     obj_from_lua->paint();
     
     lua.safe_script(code);
     
     std::cout << std::endl;
     
     return 0;
}
```

Do you expect the C++ call after defining `obj.paint()` in Lua to also change? Or are you just talking about overriding the function in Lua only?

**Comment by brinkqiang** [comment]
`
#define SOL_ALL_SAFETIES_ON 1
#include <sol/sol.hpp>

#include <iostream>

int main() {
    std::cout << "=== override-able member functions ===" << std::endl;

    struct thingy {
        thingy(){
        }

        virtual void paint() {
            std::cout << "paint in cpp" << std::endl;
        }
    };

    sol::state lua;
    lua.open_libraries(sol::lib::base);

    lua.new_usertype<thingy>("thingy",
        sol::constructors<thingy()>(),
        "paint", &thingy::paint);

    sol::string_view code = R"(
    obj = thingy.new()
    obj:paint()
    obj.paint = function (self) print("paint in lua") end
    obj:paint()
    function obj:paint () print("paint in lua2") end
    obj:paint()
    )";

    lua.safe_script(code);

    std::cout << std::endl;

    return 0;
}
`
output:
=== override-able member functions ===
paint in cpp
paint in cpp
paint in cpp

I want:
=== override-able member functions ===
paint in cpp
paint in lua
paint in lua2

**Comment by ThePhD** [comment]
I... feel like this might already work, or at least it should...?

Not sure, time to find out.

**Comment by brinkqiang** [comment]
> Related to #583 ?

Yes, I have seen the solution provided. It does not meet my needs. Using sol::function paint is intrusive. Because I use the code generation method. I don't want to modify the original implementation of the class.

**Comment by brinkqiang** [comment]
> I... feel like this might already work, or at least it should...?
> 
> Not sure, time to find out.

up

**Comment by otristan** [comment]
You can do that without modifying  the original implementation of the class.
You "just" need to bind with sol wrapped class that inherit from your component and override the virtual method to defer to a sol function or the original implementation

**Comment by brinkqiang** [comment]
> 您可以在不修改类的原始实现的情况下做到这一点。 您“只需要”绑定从组件继承的 sol 包装类并覆盖虚拟方法以遵循 sol 函数或使用模板的原始实现
> 
> ```
>  template <class T> class LuaComponentBase : public T
>   {
>   public:
>     template <typename ...Args>
>     LuaComponentBase(sol::this_state ts, Args &&...args)
>       : T(std::forward<Args>(args)...)
>     {
> }
> }
> ```

Thank you very much, you reminded me. The problem has been solved.
In the process of generating bind, generate the corresponding sol::function to bind. No need to modify the original class.
